### ISSUE TITLE
 Import endpoint should allow for any (known) work identifiers

### ISSUE THREAD

**Issue by Freso** [internetarchive/openlibrary#9472]
### TL;DR

Same as https://github.com/internetarchive/openlibrary/issues/9448 but for Works :)

### A clear and concise description of what you want to happen

Being able to include any (known to OL) identifiers for the work when importing into Open Library.

Note: This is a superset of https://github.com/internetarchive/openlibrary/issues/9471 which is only concerned with Open Library identifiers. Making a separate issue due to additional considerations (see later section).

### Expected behaviour / screenshots (ex: Figma design screenshots for UI feature)

When generating a JSON blurb for import into OL, it should be possible to provide known work identifiers in it.

### Additional Context

When importing into OL you might have a variety of identifiers available that might assist in pinpointing the correct Work (if it exists in OL). E.g., if you import from Goodreads, you will have the Goodreads work id in addition to the edition id. Right now it is not possible to provide these to the import pipeline to help with identifying authors, but it could be a great help.

### What is the proposed solution / implementation?

Changing the JSON schema to allow for identifiers for authors, perhaps something like,
```diff
diff --git a/olclient/schemata/import.schema.json b/olclient/schemata/import.schema.json
index 3f00e90..319c867 100644
--- a/olclient/schemata/import.schema.json
+++ b/olclient/schemata/import.schema.json
@@ -109,19 +109,12 @@
       ]
     },
     "identifiers": {
-      "type": "object",
-      "patternProperties": {
-        "^\\w+": { "$ref": "shared_definitions.json#/string_array" }
-      },
-      "description": "Unique identifiers used by external sites to identify a book. Used by Open Library to link offsite.",
-      "examples": [
-        {
-            "standard_ebooks": ["leo-tolstoy/what-is-art/aylmer-maude"]
-        },
-        {
-            "project_gutenberg": ["64317"]
-        }
-      ]
+      "type": "array",
+      "items": { "$ref": "#/definitions/import_identifier" }
+    },
+    "work_identifiers": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/import_identifier" }
     },
     "cover": {
       "type": "string",
@@ -132,6 +125,21 @@
     }
   },
   "definitions": {
+    "import_identifier": {
+      "type": "object",
+      "patternProperties": {
+        "^\\w+": { "$ref": "shared_definitions.json#/string_array" }
+      },
+      "description": "Unique identifiers used by external sites to identify a book, author, or work. Used by Open Library to link offsite.",
+      "examples": [
+        {
+            "standard_ebooks": ["leo-tolstoy/what-is-art/aylmer-maude"]
+        },
+        {
+            "project_gutenberg": ["64317"]
+        }
+      ]
+    },
     "import_author": {
       "type": "object",
       "additionalProperties": false,
```
and then of course have the importer pipeline actually recognise the work identifier(s) and use it for matching against existing OL works.

### Is there a precedent of this approach succeeding elsewhere?

Several MusicBrainz importer scripts use identifiers from import sources to match up identifiers in MusicBrainz. E.g., [a-tisket](https://atisket.pulsewidth.org.uk/) cross-references artist identifiers from iTunes, Deezer, and Spotify with ones known in MusicBrainz to ease the import into MusicBrainz by assigning artists to already existing ones. The [Discogs importer userscript](https://github.com/murdos/musicbrainz-userscripts/?tab=readme-ov-file#discogs_importer) does the same, but also does this for Release Groups and Labels.

Granted, the import flow for MusicBrainz is _quite_ different from Open Library, but I think it still shows how being able to look up an import source’s own identifiers can greatly help in matching against the target dataset.

### Which suggestions or requirements should be considered for how feature needs to appear or be implemented?

Some considerations:
- What happens when a work otherwise perfectly matches an existing work, but…
  - the provided identifier isn’t/identifiers aren’t already known?
    - does work get matched regardless?
      - and the identifier is/identifiers are discarded
      - and the identifier is/identifiers are added to the matched work
    - a new work gets created with the identifier(s) attached?
    - import fails
  - provided identifier(s) not known but work already has identifier(s) of the same type(s) (e.g., Goodreads id provided, but matched work already has a different Goodreads id)
  - provided identifier(s) match(es) a different work
  - provided identifier(s) match(es) different work_s_(!)
- Identifier(s) match(es) existing Work but any other provided data do not

And I’m sure there are plenty of other edge cases, but this is clearly more involved than just allowing OL ids (https://github.com/internetarchive/openlibrary/issues/9471)

### Related files

Schema: https://github.com/internetarchive/openlibrary-client/blob/master/olclient/schemata/import.schema.json (though this may be moved to this repository; see https://github.com/internetarchive/openlibrary/issues/9351 for status of this)
Import logic: https://github.com/internetarchive/openlibrary/tree/master/openlibrary/plugins/importapi

### Stakeholders

<hr />

### Instructions for Contributors

- Please [run these commands](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch) to ensure your repository is up to date **before** [creating a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) to work on this issue and **each time after** pushing code to Github, because the pre-commit bot may add commits to your PRs upstream.

**Comment by VarunBhatP** [comment]
Hi! I’m exploring OpenLibrary as a new contributor and would like to take this issue.

My plan is to reproduce the current behavior of the import endpoint, add tests for work identifiers, and propose a small fix.

Please let me know if that sounds reasonable.

**Comment by mekarpeles** [comment]
## Context & Scope Clarification

This issue is about two linked changes:

1. **Schema**: add a `work_identifiers` field to `import.schema.json` so callers can pass external identifiers (Goodreads work ID, Wikidata Q-number, etc.) alongside an import record
2. **Pipeline logic**: teach `find_matching_work()` to use those identifiers when finding the right OL Work record, rather than relying solely on normalized title-matching

It is a **superset** of #9471 (which only covers the OL `OL…W` key case). Both issues can be resolved in a single PR.

---

## Updated File Locations

The schema referenced in the issue body is the **old location** (in `openlibrary-client`). Since #9351 resolved, the schema now lives in this repo:

- **Import schema**: [`openlibrary/schemata/import.schema.json`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/schemata/import.schema.json)
- **Work-matching logic**: [`openlibrary/catalog/add_book/__init__.py` → `find_matching_work()` (line ~197)](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/add_book/__init__.py#L197)
- **Import API**: [`openlibrary/plugins/importapi/`](https://github.com/internetarchive/openlibrary/tree/master/openlibrary/plugins/importapi)

---

## Useful Precedent: Author Identifiers (#9448)

The sister issue — allowing external identifiers for **authors** on import — was closed as completed in March 2025. The author-resolution logic in [`openlibrary/catalog/add_book/load_book.py` → `find_author()`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/add_book/load_book.py) now supports `remote_ids` on import author objects:

```json
{
  "authors": [{ "name": "Charles Dickens", "remote_ids": {"librivox": "91"} }]
}
```

**Note for contributor**: `remote_ids` support was added to the *pipeline logic* but a quick check suggests it may not yet be reflected in the *import schema* (`import_author` definition in `import.schema.json`). Worth verifying before starting this issue.

The same two-part pattern applies here:
1. Add `work_identifiers` to the schema
2. Update `find_matching_work()` to try identifier-based lookup before falling back to title matching

---

## Key Decision Points (from the issue body)

The issue author has already enumerated the main edge cases. The one that needs a design decision before coding begins is **conflict priority**:

> *"provided identifier(s) match a different work"*

The author-identifier analog resolved this by: identifier match takes **priority over name matching** when present. The author is found (or created) based on identifiers, and name/date matching is a fallback. A reasonable default for works would be the same: if a `work_identifiers` entry matches an existing OL work, use it even if the title match is imperfect — but **add** the incoming identifiers to the matched work rather than discarding them.

The trickiest edge case is *identifier conflict* (the provided ID matches a *different* work than title matching would find). Recommending a conservative default: log the conflict and fall back to title matching; do not silently resolve to either.

---

## Suggested First Steps

1. Check that `remote_ids` on `import_author` is actually in the schema (may be a quick fix needed as a prerequisite)
2. Sketch the `work_identifiers` schema field (analogous to the existing `identifiers` object field, but for the work level)
3. Add identifier-based lookup to `find_matching_work()` — try each provided identifier against OL works before falling back to title normalization
4. Write tests in [`openlibrary/catalog/add_book/tests/test_add_book.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/add_book/tests/test_add_book.py) covering: identifier match, identifier conflict, identifier not found (fallback to title)

*— Richy, via PAM (Open Library's Project AI Manager)*
